payments-ui-core/src/main/java/com/stripe/android/ui/core/forms/resources/LpmRepository.kt

```diff
@@ -110,78 +113,60 @@ class LpmRepository constructor(
-                .filter { supportsPaymentMethod(it) }
-
-            if (lpmsNotParsedFromServerSpec.isNotEmpty()) {
-                val mapFromDisk: Map<String, SharedDataSpec>? =
-                    readFromDisk()
-                        ?.associateBy { it.type }
-                        ?.filterKeys { expectedLpms.contains(it) }
-                lpmInitialFormData.putAll(
-                    lpmsNotParsedFromServerSpec
-                        .mapNotNull { mapFromDisk?.get(it) }
-                        .mapNotNull { convertToSupportedPaymentMethod(it) }
-                        .associateBy { it.code }
-                )
-                mapFromDisk
-                    ?.mapValues { it.value.nextActionSpec.transform() }
-                    ?.let {
-                        lpmPostConfirmData.update(
-                            it
-                        )
-                    }
+        val countDownLatch = CountDownLatch(1)
+        serverInitializedLatch = countDownLatch
+        val expectedLpms = stripeIntent.paymentMethodTypes
+
+        serverSpecLoadingState = ServerSpecState.NoServerSpec(serverLpmSpecs)
+        if (!serverLpmSpecs.isNullOrEmpty()) {
+            serverSpecLoadingState = ServerSpecState.ServerNotParsed(serverLpmSpecs)
+            val serverLpmObjects = lpmSerializer.deserializeList(serverLpmSpecs)
+            if (serverLpmObjects.isNotEmpty()) {
+                serverSpecLoadingState = ServerSpecState.ServerParsed(serverLpmSpecs)
             }
+            update(stripeIntent, serverLpmObjects)
+        }
 
-            serverInitializedLatch.countDown()
+        // If the server does not return specs, or they are not parsed successfully
+        // we will use the LPM on disk if found
+        val lpmsNotParsedFromServerSpec = expectedLpms
+            .filter { !lpmInitialFormData.containsKey(it) }
+            .filter { supportsPaymentMethod(it) }
+
+        if (lpmsNotParsedFromServerSpec.isNotEmpty()) {
+            val mapFromDisk: Map<String, SharedDataSpec>? =
+                readFromDisk()
+                    ?.associateBy { it.type }
+                    ?.filterKeys { expectedLpms.contains(it) }
+            lpmInitialFormData.putAll(
+                lpmsNotParsedFromServerSpec
+                    .mapNotNull { mapFromDisk?.get(it) }
+                    .mapNotNull { convertToSupportedPaymentMethod(stripeIntent, it) }
+                    .associateBy { it.code }
+            )
+            mapFromDisk
+                ?.mapValues { it.value.nextActionSpec.transform() }
+                ?.let {
+                    lpmPostConfirmData.update(
+                        it
+                    )
+                }
         }
+
+        countDownLatch.countDown()
     }
 
     @VisibleForTesting
-    fun updateFromDisk() {
-        update(readFromDisk())
+    fun updateFromDisk(stripeIntent: StripeIntent) {
+        update(stripeIntent, readFromDisk())
     }
```

Comment:
Nit: Can we get rid of this too?

Reply:
I think this is only used in tests. So we could get rid of it. But we'd either have to make readFromDisk public, or copy it's implementation for tests.